File: src/codes/duplicate-logic/ts/rfc_decorator.ts
// コンポーネントインターフェース
interface Exporter {
  export(data: string): void;
}

// 具象コンポーネント
class PdfExporter implements Exporter {
  export(data: string): void {
    console.log(`PDF出力: ${data}`);
  }
}

class CsvExporter implements Exporter {
  export(data: string): void {
    console.log(`CSV出力: ${data}`);
  }
}

// デコレータ（共通機能）
class LoggingExporter implements Exporter {
  constructor(private wrapped: Exporter) {}

  export(data: string): void {
    console.log("開始ログ");
    this.wrapped.export(data);
    console.log("完了ログ");
  }
}

// 利用例
const pdf = new LoggingExporter(new PdfExporter());
pdf.export("帳票データ");

const csv = new LoggingExporter(new CsvExporter());
csv.export("ユーザー一覧");

File: src/codes/duplicate-logic/py/rfc_decorator.py
from abc import ABC, abstractmethod

class Exporter(ABC):
    @abstractmethod
    def export(self, data: str):
        pass

class PdfExporter(Exporter):
    def export(self, data: str):
        print(f"PDF出力: {data}")

class CsvExporter(Exporter):
    def export(self, data: str):
        print(f"CSV出力: {data}")

class LoggingExporter(Exporter):
    def __init__(self, wrapped: Exporter):
        self._wrapped = wrapped

    def export(self, data: str):
        print("開始ログ")
        self._wrapped.export(data)
        print("完了ログ")

# 利用例
pdf = LoggingExporter(PdfExporter())
pdf.export("帳票データ")

csv = LoggingExporter(CsvExporter())
csv.export("ユーザー一覧")
